{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tool functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from gensim.corpora.dictionary import Dictionary\n",
    "from gensim.models import LdaModel\n",
    "from gensim.parsing.preprocessing import preprocess_string\n",
    "from gensim.models import TfidfModel\n",
    "\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_sentences_from_text(text_lines, min_line_len=20, min_sen_len=15):\n",
    "    fulltext = \" \".join([line for line in list(text_lines) if len([l for l in line.split(\" \") if len(l) > 0]) > min_line_len])\n",
    "    sens = filter(lambda sen: len(sen) >= min_sen_len, fulltext.split(\".\"))\n",
    "    return \". \".join(sens)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_texts_from_dir(texts_dir):\n",
    "    txt_files = os.listdir(texts_dir)\n",
    "    txt_files = [os.path.join(texts_dir, txt) for txt in txt_files]\n",
    "    texts = dict()\n",
    "    for txt_f in list(filter(lambda path: path.endswith(\".txt\"), txt_files)):\n",
    "        try:\n",
    "            text = filter_sentences_from_text(open(txt_f, \"r\").readlines())\n",
    "            texts[txt_f] = preprocess_string(text)\n",
    "        except UnicodeDecodeError:\n",
    "            print(\"Utf-8 decode error on %s\" % txt_f)\n",
    "            continue\n",
    "    return texts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_texts(texts_dir):\n",
    "    txt_files = os.listdir(texts_dir)\n",
    "    txt_files = [os.path.join(texts_dir, txt) for txt in txt_files]\n",
    "    for txt_f in list(filter(lambda path: path.endswith(\".txt\"), txt_files)):\n",
    "        yield open(txt_f, \"r\").readlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['\\ufeffSkip to main content \\n',\n",
       " ' \\n',\n",
       " 'Thank you for visiting nature.com. You are using a browser version with limited support for CSS. To obtain the best experience, we recommend you use a more up to date browser (or turn off compatibility mode in Internet Explorer). In the meantime, to ensure continued support, we are displaying the site without styles and JavaScript.\\n',\n",
       " '    1. nature\\n',\n",
       " '    2. translational psychiatry\\n',\n",
       " '    3. original article\\n',\n",
       " '    4. article\\n',\n",
       " 'Menu \\n',\n",
       " '\\n',\n",
       " 'Search E-alert Submit My Account Login \\n',\n",
       " '    • PDF\\n',\n",
       " 'Polygenic loading for major depression is associated with specific medical comorbidity\\n',\n",
       " 'Close menu\\n',\n",
       " 'Search\\n',\n",
       " 'Search nature.com  \\n',\n",
       " 'advanced \\n',\n",
       " 'Search \\n',\n",
       " 'Close menuClose menu\\n',\n",
       " 'Menu\\n',\n",
       " 'Translational Psychiatry \\n',\n",
       " '    • Browse Articles \\n',\n",
       " '    • Browse Collections \\n',\n",
       " '    • Focuses \\n',\n",
       " '    • About the Journal \\n',\n",
       " '    • Open Access \\n',\n",
       " '    • About the Editors \\n',\n",
       " '    • For Authors & Referees \\n',\n",
       " '    • Contact \\n',\n",
       " '    • For Advertisers \\n',\n",
       " '    • Press Releases \\n',\n",
       " 'Journals A-Z Journals A-Z \\n',\n",
       " '    • Nature \\n',\n",
       " '    • Nature Communications \\n',\n",
       " '    • Nature Protocols \\n',\n",
       " '    • Review journals \\n',\n",
       " '    • Scientific Reports \\n',\n",
       " '    • View all journals \\n',\n",
       " 'All Subjects All Subjects \\n',\n",
       " '    • Biological Sciences \\n',\n",
       " '    • Earth & Environmental Sciences \\n',\n",
       " '    • Health Sciences \\n',\n",
       " '    • Physical Sciences \\n',\n",
       " '    • Scientific Community & Society \\n',\n",
       " '    • View all subjects \\n',\n",
       " 'nature.com nature.com \\n',\n",
       " '    • Subscriptions \\n',\n",
       " '    • Authors & Referees \\n',\n",
       " '    • Librarians \\n',\n",
       " '    • Advertisers \\n',\n",
       " '    • Press \\n',\n",
       " '    • About \\n',\n",
       " 'Close menu\\n',\n",
       " 'Original Article | Open | Published: 19 September 2017 \\n',\n",
       " 'Polygenic loading for major depression is associated with specific medical comorbidity\\n',\n",
       " '    • T H McCoy1\\n',\n",
       " '    • , V M Castro1, 2\\n',\n",
       " '    • , L Snapper1\\n',\n",
       " '    • , K Hart1\\n',\n",
       " '    • , J L Januzzi3\\n',\n",
       " '    • , J C Huffman4\\n',\n",
       " '    • […]\\n',\n",
       " '    • \\xa0& R H Perlis1\\n',\n",
       " '    • - Show fewer authors\\n',\n",
       " 'Translational Psychiatry volume 7, page e1238 (2017) | Download Citation\\n',\n",
       " 'AbstractAbstract\\n',\n",
       " \"Major depressive disorder frequently co-occurs with medical disorders, raising the possibility of shared genetic liability. Recent identification of 15 novel genetic loci associated with depression allows direct investigation of this question. In cohorts of individuals participating in biobanks at two academic medical centers, we calculated polygenic loading for risk loci reported to be associated with depression. We then examined the association between such loading and 50 groups of clinical diagnoses, or topics, drawn from these patients' electronic health records, determined using a novel application of latent Dirichilet allocation. Three topics showed experiment-wide association with the depression liability score; these included diagnostic groups representing greater prevalence of mood and anxiety disorders, greater prevalence of cardiac ischemia, and a decreased prevalence of heart failure. The latter two associations persisted even among individuals with no mood disorder diagnosis. This application of a novel method for grouping related diagnoses in biobanks indicate shared genetic risk for depression and cardiac disease, with a pattern suggesting greater ischemic risk and diminished heart failure risk.\\n\",\n",
       " 'IntroductionIntroduction\\n',\n",
       " 'Major depression has been associated with a host of non-psychiatric comorbidities, ranging from cardiovascular disease to autoimmune disorders. The impact of co-occurrence is often profound and bi-directional—that is, outcomes of each disorder tend to be worse in the presence of the other,1, 2, 3, 4, 5 despite the availability of multiple efficacious treatments.6 Thus, understanding the mechanism of such co-occurrence has both scientific and clinical relevance.\\n',\n",
       " 'This co-occurrence has a range of possible explanations. It is possible that depression increases risk for some disorders, either directly (for example, via changes in cortisol or immunosuppression) or indirectly (for example, via changes in health behaviors).7, 8 Conversely, the presence of a medical disorder can represent a stressor increasing risk for depression.9 Yet, a third model posits a shared liability—i.e., the same factors that increase risk for one disorder may increase risk for the other.10 Multiple twin or family investigations support this shared liability, at least for cardiovascular disease.11, 12, 13 In reality, it is likely that all three mechanisms play a role in co-occurrence of some disorders, but confirming the presence of at least one mechanism would represent a key first step in understanding these relationships.14\\n',\n",
       " 'Multiple novel genetic loci associated with major depression in individuals of Northern European ancestry at a genome-wide level of significance have recently been reported.15 These newly reported loci allow for direct examination of the possibility that aggregated genetic liability for depression, in addition to depression itself, is associated with risk for disorders other than depression. The present study aims to test the association between such liability and medical comorbidity, using a novel method for deriving this comorbidity. Specifically, we drew on these newly reported genetic associations to construct estimates of common-variant genetic loading in the large medical biobanks from two academic medical centers. We then tested these depression loading measures for association with empirically derived groups of diagnoses, or topics, that tend to co-occur with each other.\\n',\n",
       " 'Materials and methodsMaterials and methods\\n',\n",
       " 'Clinical phenotype derivation\\n',\n",
       " 'Standard phenome-wide association studies (PheWAS) test all diagnostic codes against all predictors—typically individual risk variants or genome-wide common variants.16 This approach risks either inflation of type I error (by testing 1500+ diagnostic codes), or type II error (by correcting for 1500+ diagnostic codes). Further, it fails to take into account correlation between individual coded diagnoses, and the highly variable reliability of many such codes.\\n',\n",
       " \"As an alternative, we developed a method that applies latent Dirichilet allocation (LDA) to reduce categorical diagnostic ontologies to a finite set of topics on which these codes load.17, 18 This form of unsupervised machine learning has most commonly been applied in natural language processing to capture the topics expressed in documents; it presumes that individual tokens or terms (here, diagnostic codes) reflect an underlying topic, and that the record of an individual patient reflects a combination of latent topics. Figure 1 depicts the process of conceptualizing patient medical records as a 'bag' of observed diagnostic code counts from which unobserved latent topics are inferred using LDA. Thereafter, the inferred topics are treated as the phenotype in analysis.\\n\",\n",
       " 'Figure 1\\n',\n",
       " '\\n',\n",
       " 'Illustration of the process of topic modeling as applied in this study.\\n',\n",
       " 'Full size image\\n',\n",
       " \"Here we extracted all ICD-9 diagnostic codes for biobank participants from the inpatient and outpatient electronic medical records of Massachusetts General Hospital and Brigham and Women's Hospital and grouped them into 1667 PheWAS disease categories.19 All participants had signed written informed consent for biobank participation, including consent for release of deidentified data under a Data Use Agreement to qualified investigators, as approved by the hospitals' Institutional Review Board. We eliminated PheWAS codes occurring in <1% of individuals or more than 99% of individuals in the first cohort, leaving 508 codes for topic model construction. We then trained a 50-topic model, and scored cohorts one and two. (The decision a priori to select 50 topics is discussed further below; the optimal number of targets remains an area of research in unsupervised learning20, 21, 22, 23). The LDA was performed using the Gensim implementation.24, 25 Importantly LDA allows for the possibility of all codes with respect to all topics. The distinction between topics is in the expected probabilities of each code. As such, we focus on the most probable diagnostic codes given each topic as a means of interpreting the topics. When individual topics are mentioned in the text they are named using the most strongly loaded code with the suffix ‘++’ to indicate that a topic comprises many codes, each contributing to membership in that topic.\\n\",\n",
       " 'Molecular methods\\n',\n",
       " 'All subjects were genotyped using either the Illumina MEGA (cohort 1, n=4931) or the Illumina MEGA-EX (cohort 2, n=4428) array (Illumina; San Diego, CA, USA). Each cohort was cleaned, imputed, and analyzed separately to minimize batch effects. We retained subjects with genotyping call rates exceeding 99% and no evidence of relatedness based on identity by descent (IBD).26 We likewise retained any SNPs with call rate of 95% or greater, and Hardy–Weinberg equilibrium P-value>1 × 10−10. Genotypes were next imputed using the Michigan Imputation Server implementing Minimac3, based on all population subsets from 1000 Genomes Phase 3 v5 as reference panel.27, 28, 29 Phasing of haplotypes used SHAPEIT.30\\n',\n",
       " 'We generated principal components, as implemented in PLINK 1.9, to identify the first 10 components of the variance-standardized relationship matrix among genotyped SNPs in each cohort.31 After overlaying HapMap populations, only those individuals falling within the Northern European cluster were included in subsequent analysis.\\n',\n",
       " 'Analysis\\n',\n",
       " \"We generated polygenic risk scores (PRS), estimates of polygenic loading for MDD, using seven tranches of SNPs (5 × 10−8, 1 × 10−7, 1 × 10−6, 1 × 10−5, 1 × 10−4, 1 × 10−3 and 1 × 10−2, S1–S7) drawn from our prior publication reporting 15 genome-wide associations with depression.15 The value for each P-value tranche represents the maximum P-value that is included in that tranche. This list was linkage-disequilibrium pruned using the 'clump' function as implemented in PLINK 1.9, with a 250\\u2009kb window and minimum r2 set at 0.5 by default.32\\n\",\n",
       " 'We used linear regression to examine association between depression polygenic score and each of the 50 topics. We fit both unadjusted models and models incorporating the first 10 MDS components, and present the meta-analyzed result of the two genotyping cohorts.\\n',\n",
       " 'As this analysis was intended as a hypothesis-generating effort, we present uncorrected P-values in all results. For purposes of interpretation, Bonferroni correction for 50 topics would require a P-value of 0.05/50, or 0.001, for significance. The seven PRS tranches are correlated so do not represent seven independent tests per phenotype; since the average r between them is ~0.65, we consider a fully corrected P-value threshold for significance to be 0.00042.33, 34 To further elucidate the statistical significance of any associations identified we utilized permutation to calculate empirical P-values as well as an experiment-wide P-value. To do so, we randomized the relationship between the topics (phenotype) and the MDS adjusted PRS (genotype) and calculated association in the full cohort between all tranches and all topics 100\\u2009000 times under this simulated null.\\n',\n",
       " 'ResultsResults\\n',\n",
       " 'In the first genotyping cohort, there were 3728 individuals, including 2165 females (58.1%) and 997 individuals (26.7%) with a mood disorder. Cohort 2 included 2712 individuals, including 712 (49.2%) females and 779 individuals (28.7%) with a mood disorder. Mean age in cohort 1 was 57.6 (s.d. 16.8) years; mean age in cohort 2 was 62.2 (s.d. 15.9) years.\\n',\n",
       " 'Figure 2 illustrates the distribution of associations with PRS by topic and minimum P-value (that is, PRS threshold yielding strongest evidence of association). For consistency with other data clustering methods, topics are named according to predominant terms, adding the suffix ‘++’ to indicate that topics may contain overlapping terms as well as apparently unrelated terms. Three topics—mood disorder++ (03), heart failure++ (21) and cardiac ischemia++ (27)—met an experiment-wide threshold for association (Table 1). PheWAS diagnostic codes contributing to these topics are listed in Table 2, ranked by order of contribution (that is, weighting of each code for a given topic, from greatest to least). (Supplementary Table 1 reports all topics by all PRS tranches, sorted by association P-value). Permuted topic level associations matched those of the primary analysis (Supplementary Figure 1). In experiment-wide permutation analysis the number of significant topics was itself statistically significant (permuted P=0.02).\\n',\n",
       " 'Figure 2\\n',\n",
       " '\\n',\n",
       " 'Distribution of associations of polygenic risk scores (PRS) by topic (x axis) and minimum P-value across seven PRS tranches (yaxis).\\n',\n",
       " 'Full size image\\n',\n",
       " 'Table 1: Topics associated with PRS at an experiment-wide threshold for significance in primary analysis (left); follow-up analysis of primary associations (right)\\n',\n",
       " 'Full size table\\n',\n",
       " 'Table 2: List of top 20 PheWAS codes contributing to topics associated with PRS\\n',\n",
       " 'Full size table\\n',\n",
       " 'In secondary analysis, we excluded any individuals with a mood disorder PheWAS code (n=1776) and repeated these analyses. Association P-values are indicated in Table 1 (right), and visualized in a heat map in Figure 3. As anticipated, the mood disorder topic (03) was no longer significant, suggesting that the additional codes in that topic do not contribute meaningfully to association; the two cardiac topics demonstrate persistent association.\\n',\n",
       " 'Figure 3\\n',\n",
       " '\\n',\n",
       " 'Heat map of association for all topics, by polygenic risk scores (PRS) tranche excluding depression cases.\\n',\n",
       " 'Full size image\\n',\n",
       " 'DiscussionDiscussion\\n',\n",
       " 'In this analysis of electronic health record data from 6440 individuals of Northern European ancestry, we identified three sets of diagnoses (topics) associated with PRS for depression at an experiment-wide significance threshold. One of these encompasses mood disorders, and can be considered a positive control or indicator of assay sensitivity—though notably, it also includes related codes (adjustment disorder, tobacco use, and anxiety) that may reflect true pleiotropy or simply differences in the way mood and anxiety symptoms are coded. The other two reflect different elements of cardiac pathology. That cardiac pathology surfaces from this unsupervised machine learning approach is reassuringly face valid given the extensive literature relating cardiac and mood disorders; however, cardiac pathology is subdivided into distinct associations.1, 2, 3, 4 Cardiac ischemia++ (topic 27) largely captures acute coronary syndromes, including risk factors for such syndromes (See Table 2 for the codes most strongly associated with this topic). The data suggest that MDD loading is associated with greater risk for these acute syndromes (Table 1). Heart failure++ (topic 21) also reflects cardiovascular pathology but appears to reflect a more chronic disease course, including heart failure but not limited to cardiac disease (Table 2). Depressive genetic loading is inversely associated with heart failure++ (Table 1). This differential result underscores the complexity of the relationship between depression and cardiovascular disease.35, 36, 37 Importantly, evidence of association persists even when individuals with a mood disorder are excluded, suggesting that the observed associations are not simply consequences of a mood disorder diagnosis.\\n',\n",
       " 'The relationship between depressive disorder and acute coronary syndromes is well described; however, the mechanism remains unclear.38, 39, 40, 41, 42 Multiple investigations using twin- and family-based designs have found evidence of shared heritable liability between MDD and cardiac disease.11, 12, 13 One study suggests the complexity of this relationship, with the extent of liability varying with sex and age.13 Further complicating efforts to understand comorbidity is the observation that depression is associated with increased deleterious behavior such as tobacco use and less exercise - notably tobacco use, for which risk is likely to be heritable, contributes to both a mood++ and a cardiovascular++ topic in our results. On the other hand, in prior studies individuals with major depression exhibit greater platelet reactivity and prevalent endothelial dysfunction relative to non-depressed subjects, which is a candidate pathway for depression increasing cardiac risk directly.43\\n',\n",
       " 'Surprisingly, in the present study depression loading is inversely associated with more chronic forms of cardiac disease (heart failure++ topic 21). Although it is speculative, the divergences between acute and chronic cardiac pathology may reflect the tension between genetic and behavioral determinates of health. In the chronic phases of cardiac disease, personality is an important determinate of outcome.44, 45, 46 Alternatively, those with heart failure and comorbid depression might in fact have substantially higher mortality risk, obscuring a chronic link between the two diagnoses due to attrition.\\n',\n",
       " 'In interpreting these associations, it is important to recognize that the methodology itself identifies co-occurring codes; the application of names to the underlying (that is, latent) concepts requires human intervention. Moreover, codes are associated, even if only negligibly, with all topics. This characteristic represents both a feature and a limitation of the methodology: it does not require manual curation of topics, and thus can discover relationships in data not specified a priori; on the other hand, it requires interpretation of results that may not be intuitive. Concretely, we would not have expected acute and chronic cardiac disease to primarily occur in distinct topics. This surprise bespeaks the hypothesis-generating potential of unsupervised machine learning, which still requires interpretation and follow-up. For ease of reading, by convention we name each topic by the most strongly-loading code, but again emphasize that a topic is not simply a single code and note that visualization and interpretation of probabilistic topics remain an area of active investigation in machine learning.47 Conversely, for testing specific hypotheses about a particular set of diagnoses, merely testing a curated code or set of codes would be most interpretable, but preclude discovery of new or unanticipated relationships.\\n',\n",
       " 'Latent Dirichilet allocation has been applied extensively in natural language processing to identify topics reflected in blocks of text; however, to our knowledge, it has not been utilized for genetic investigation. The present study suggests the utility of this method for understanding the relationship between psychiatric and comorbid disorders, and we hope this study will prompt further investigations in larger biobanks and in using other polygenic risk measures. Still, an important caveat to this work is the inability to determine causation. Specifically, we cannot distinguish between the case of shared genetic liability for depression and comorbidities, and depression ‘causing’ the observed comorbidity. However, we are able to render the third possibility (the comorbidity causes the depression) less likely, by conditioning on an aggregate measure of genetic depression liability. A further caveat is the likelihood that other parameters would yield differing results—i.e., the application of 10 or 100 topics. Here, we selected 50 as a plausible number of disease groups a priori, and based on prior work with LDA in other contexts. Understanding the extent to which other, nonlinear means of weighting diagnoses, or specifying other numbers of topics, would yield differing results merits further study.\\n',\n",
       " 'Taken together, our results suggest the complexity of the relationship between mood disorders and comorbid somatic illness, and indicate specific groups of diagnoses that may travel along with genetic risk for major depression. They further illustrate the application of a novel approach to aggregating diagnoses, applicable to any large clinical data set, which may be more tolerant of heterogeneity in diagnostic codes and sensitive to groups of diagnoses that travel together. At the same time, the observed association with mood disorders may be considered further replication of the previous report of depression liability genes.15 The application of topic modeling therefore appears to be a promising strategy.\\n',\n",
       " 'ReferencesReferences\\n',\n",
       " '    1. 1.\\n',\n",
       " '       Whooley MA, de Jonge P, Vittinghoff E, Otte C, Moos R, Carney RM et al. Depressive symptoms, health behaviors, and risk of cardiovascular events in patients with coronary heart disease. JAMA 2008; 300: 2379–2388.\\n',\n",
       " '        ◦ Show context for reference 1 \\n',\n",
       " '            ▪ CAS\\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    2. 2.\\n',\n",
       " '       Whooley MA. Depression and cardiovascular disease: healing the broken-hearted. JAMA 2006; 295: 2874–2881.\\n',\n",
       " '        ◦ Show context for reference 2 \\n',\n",
       " '            ▪ CAS\\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    3. 3.\\n',\n",
       " '       Ducat L, Philipson LH, Anderson BJ. The mental health comorbidities of diabetes. JAMA 2014; 312: 691–692.\\n',\n",
       " '        ◦ Show context for reference 3 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    4. 4.\\n',\n",
       " '       Golden SH, Lazo M, Carnethon M, Bertoni AG, Schreiner PJ, Diez Roux AV et al. Examining a bidirectional association between depressive symptoms and diabetes. JAMA 2008; 299: 2751–2759.\\n',\n",
       " '        ◦ Show context for reference 4 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    5. 5.\\n',\n",
       " '       Cohen BE, Martires KJ, Ho RS. Psoriasis and the risk of depression in the US population: National Health and Nutrition Examination Survey 2009–2012. JAMA Dermatol 2016; 152: 73–79.\\n',\n",
       " '        ◦ Show context for reference 5 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    6. 6.\\n',\n",
       " '       Girardi P, Pompili M, Innamorati M, Mancini M, Serafini G, Mazzarini L et al. Duloxetine in acute major depression: review of comparisons to placebo and standard antidepressants using dissimilar methods. Hum Psychopharmacol 2009; 24: 177–190.\\n',\n",
       " '        ◦ Show context for reference 6 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    7. 7.\\n',\n",
       " '       Blume J, Douglas SD, Evans DL. Immune suppression and immune activation in depression. Brain Behav Immun 2011; 25: 221–229.\\n',\n",
       " '        ◦ Show context for reference 7 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    8. 8.\\n',\n",
       " '       Penninx BW. Depression and cardiovascular disease: epidemiological evidence on their linking mechanisms. Neurosci Biobehav Rev 2017; 74(Part B): 277–286.\\n',\n",
       " '        ◦ Show context for reference 8 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    9. 9.\\n',\n",
       " '       Tang P-L, Wang H-H, Chou F-H. A systematic review and meta-analysis of demoralization and depression in patients with cancer. Psychosomatics 2015; 56: 634–643.\\n',\n",
       " '        ◦ Show context for reference 9 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    10. 10.\\n',\n",
       " '       Cheng J, Zhang J, Lu C, Wang L. Using optogenetics to translate the “inflammatory dialogue” between heart and brain in the context of stress. Neurosci Bull 2012; 28: 435–448.\\n',\n",
       " '        ◦ Show context for reference 10 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    11. 11.\\n',\n",
       " '       van Hecke O, Hocking LJ, Torrance N, Campbell A, Padmanabhan S, Porteous DJ et al. Chronic pain, depression and cardiovascular disease linked through a shared genetic predisposition: Analysis of a family-based cohort and twin study. PLoS One 2017; 12: e0170653.\\n',\n",
       " '        ◦ Show context for reference 11 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    12. 12.\\n',\n",
       " '       Scherrer JF, Xian H, Bucholz KK, Eisen SA, Lyons MJ, Goldberg J et al. A twin study of depression symptoms, hypertension, and heart disease in middle-aged men. Psychosom Med 2003; 65: 548–557.\\n',\n",
       " '        ◦ Show context for reference 12 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    13. 13.\\n',\n",
       " '       Kendler KS, Gardner CO, Fiske A, Gatz M. Major depression and coronary artery disease in the swedish twin registry: Phenotypic, genetic, and environmental sources of comorbidity. Arch Gen Psychiatry 2009; 66: 857–863.\\n',\n",
       " '        ◦ Show context for reference 13 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    14. 14.\\n',\n",
       " '       Smith IG, Parker G, Rourke P, Cvejic E, Vollmer-Conna U. Acute coronary syndrome and depression: A review of shared pathophysiological pathways. Aust N Z J Psychiatry 2015; 49: 994–1005.\\n',\n",
       " '        ◦ Show context for reference 14 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    15. 15.\\n',\n",
       " '       Hyde CL, Nagle MW, Tian C, Chen X, Paciga SA, Wendland JR et al. Identification of 15 genetic loci associated with risk of major depression in individuals of European descent. Nat Genet 2016; 48: 1031–1036.\\n',\n",
       " '        ◦ Show context for reference 15 \\n',\n",
       " '            ▪ CAS\\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    16. 16.\\n',\n",
       " '       Antony A, Tang YS, Khan RA, Biju MP, Xiao X, Li QJ et al. Translational upregulation of folate receptors is mediated by homocysteine via RNA-heterogeneous nuclear ribonucleoprotein E1 interactions. J Clin Invest 2004; 113: 285–301.\\n',\n",
       " '        ◦ Show context for reference 16 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    17. 17.\\n',\n",
       " '       Blei DM, Ng AY, Jordan MI. Latent dirichlet allocation. J Machine Learn Res 2003; 3: 993–1022.\\n',\n",
       " '        ◦ Show context for reference 17 \\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    18. 18.\\n',\n",
       " '       Blei DM. Surveying a suite of algorithms that offer a solution to managing large document archives. Commun ACM 2012; 55: 77–84.\\n',\n",
       " '        ◦ Show context for reference 18 \\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    19. 19.\\n',\n",
       " '       Denny JC, Ritchie MD, Basford MA, Pulley JM, Bastarache L, Brown-Gentry K et al. PheWAS: demonstrating the feasibility of a phenome-wide scan to discover gene–disease associations. Bioinformatics 2010; 26: 1205–1210.\\n',\n",
       " '        ◦ Show context for reference 19 \\n',\n",
       " '            ▪ CAS\\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    20. 20.\\n',\n",
       " '       Arun R, Suresh V, Veni Madhavan C, Narasimha Murthy M. On finding the natural number of topics with latent dirichlet allocation: Some observations.In:Advance in Knowledge Discovery and Data Mining, Hyderabad, India,2010,pp391–402.\\n',\n",
       " '        ◦ Show context for reference 20 \\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    21. 21.\\n',\n",
       " '       Cao J, Xia T, Li J, Zhang Y, Tang S. A density-based method for adaptive LDA model selection. Neurocomputing 2009; 72: 1775–1781.\\n',\n",
       " '        ◦ Show context for reference 21 \\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    22. 22.\\n',\n",
       " '       Deveaud R, SanJuan E, Bellot P. Accurate and effective latent concept modeling for ad hoc information retrieval. Document numérique 2014; 17: 61–84.\\n',\n",
       " '        ◦ Show context for reference 22 \\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    23. 23.\\n',\n",
       " '       Griffiths TL, Steyvers M. Finding scientific topics. Proc Natl Acad Sci USA 2004; 101(suppl 1): 5228–5235.\\n',\n",
       " '        ◦ Show context for reference 23 \\n',\n",
       " '            ▪ CAS\\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    24. 24.\\n',\n",
       " '       Rehurek R, Sojka PSoftware framework for topic modelling with large corpora. In:Proceedings of the LREC 2010 Workshop on New Challenges for NLP Frameworks. Citeseer: Valletta, Malta, 2010,pp45–50.\\n',\n",
       " '        ◦ Show context for reference 24 \\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    25. 25.\\n',\n",
       " '       Hoffman M, Blei D, Bach FOnline learning for latent dirichlet allocation In: Advances in Neural Information Processing Systems 2010,pp856–864.\\n',\n",
       " '        ◦ Show context for reference 25 \\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    26. 26.\\n',\n",
       " \"       Henn BM, Hon L, Macpherson JM, Eriksson N, Saxonov S, Pe'er I et al. Cryptic distant relatives are common in both isolated and cosmopolitan genetic samples. PLoS ONE 2012; 7: e34267.\\n\",\n",
       " '        ◦ Show context for reference 26 \\n',\n",
       " '            ▪ CAS\\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    27. 27.\\n',\n",
       " '       Fuchsberger C, Abecasis GR, Hinds DAMinimac2: faster genotype imputation. Bioinformatics 2015; 31: 782–784.\\n',\n",
       " '        ◦ Show context for reference 27 \\n',\n",
       " '            ▪ CAS\\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    28. 28.\\n',\n",
       " '       Fuchsberger C, Forer L, Schoenherr S, Das S, Abecasis G.Michigan Imputation Server. Available at https://imputationserver.sph.umich.edu/index.html#!pages/home(accessed on 21 December 2016.)\\n',\n",
       " '        ◦ Show context for reference 28 \\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    29. 29.\\n',\n",
       " '       Minimac3. Center for statistical genetics. Available athttp://genome.sph.umich.edu/wiki/Minimac3(accessed on 21 Deecember 2016.)\\n',\n",
       " '        ◦ Show context for reference 29 \\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    30. 30.\\n',\n",
       " '       Delaneau O, Marchini J, Zagury J-F. A linear complexity phasing method for thousands of genomes. Nature methods 2012; 9: 179–181.\\n',\n",
       " '        ◦ Show context for reference 30 \\n',\n",
       " '            ▪ CAS\\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    31. 31.\\n',\n",
       " '       Chang CC, Chow CC, Tellier LC, Vattikuti S, Purcell SM, Lee JJ. Second-generation PLINK: rising to the challenge of larger and richer datasets. GigaScience 2015; 4: 7.\\n',\n",
       " '        ◦ Show context for reference 31 \\n',\n",
       " '            ▪ CAS\\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    32. 32.\\n',\n",
       " '       Purcell S, Neale B, Todd-Brown K, Thomas L, Ferreira MA, Bender D et al. PLINK: a tool set for whole-genome association and population-based linkage analyses. Am J Hum Genet 2007; 81: 559–575.\\n',\n",
       " '        ◦ Show context for reference 32 \\n',\n",
       " '            ▪ CAS\\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    33. 33.\\n',\n",
       " '       Sankoh AJ, Huque MF, Dubey SD. Some comments on frequently used multiple endpoint adjustment methods in clinical trials. Stat Med 1997; 16: 2529–2542.\\n',\n",
       " '        ◦ Show context for reference 33 \\n',\n",
       " '            ▪ CAS\\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    34. 34.\\n',\n",
       " \"       Perneger TV. What's wrong with Bonferroni adjustments. BMJ 1998; 316: 1236–1238.\\n\",\n",
       " '        ◦ Show context for reference 34 \\n',\n",
       " '            ▪ CAS\\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    35. 35.\\n',\n",
       " '       Angermann CE, Gelbrich G, Störk S, Gunold H, Edelmann F, Wachter R et al. Effect of escitalopram on All-cause mortality and hospitalization in patients with heart failure and depression: The MOOD-HF randomized clinical trial. JAMA 2016; 315: 2683.\\n',\n",
       " '        ◦ Show context for reference 35 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    36. 36.\\n',\n",
       " '       Angermann C, Deckert J, Ertl G. Escitalopram and outcomes among patients with depression and heart failure-reply. JAMA 2016; 316: 1494.\\n',\n",
       " '        ◦ Show context for reference 36 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    37. 37.\\n',\n",
       " '       Ströhle A, Rieckmann N. Escitalopram and outcomes among patients with depression and heart failure. JAMA 2016; 316: 1494.\\n',\n",
       " '        ◦ Show context for reference 37 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    38. 38.\\n',\n",
       " '       Carney RM, Rich MW, Freedland KE, Saini J, Simeone C, Clark K. Major depressive disorder predicts cardiac events in patients with coronary artery disease. Psychosom Med 1988; 50: 627–633.\\n',\n",
       " '        ◦ Show context for reference 38 \\n',\n",
       " '            ▪ CAS\\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    39. 39.\\n',\n",
       " '       Van Melle JP, De Jonge P, Spijkerman TA, Tijssen JG, Ormel J, Van Veldhuisen DJ et al. Prognostic association of depression following myocardial infarction with mortality and cardiovascular events: a meta-analysis. Psychosom Med 2004; 66: 814–822.\\n',\n",
       " '        ◦ Show context for reference 39 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    40. 40.\\n',\n",
       " '       Lespérance F, Frasure-Smith N, Talajic M. Major depression before and after myocardial infarction: its nature and consequences. Psychosom Med 1996; 58: 99–110.\\n',\n",
       " '        ◦ Show context for reference 40 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    41. 41.\\n',\n",
       " '       Rumsfeld JS, Jones PG, Whooley MA, Sullivan MD, Pitt B, Weintraub WS et al. Depression predicts mortality and hospitalization in patients with myocardial infarction complicated by heart failure. Am Heart J 2005; 150: 961–967.\\n',\n",
       " '        ◦ Show context for reference 41 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    42. 42.\\n',\n",
       " '       Williams M. Platelets and depression in cardiovascular disease: a brief review of the current literature. World J Psychiatr 2012; 2: 114–123.\\n',\n",
       " '        ◦ Show context for reference 42 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    43. 43.\\n',\n",
       " '       Can MM, Guler G, Guler E, Ozveren O, Turan B, DiNicolantinio JJ et al. Enhanced platelet reactivity in pediatric depression: an observational study. Blood Coagul Fibrinolysis 2015; 26: 731–735.\\n',\n",
       " '        ◦ Show context for reference 43 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    44. 44.\\n',\n",
       " '       Rozanski A, Blumenthal JA, Kaplan J. Impact of psychological factors on the pathogenesis of cardiovascular disease and implications for therapy. Circulation 1999; 99: 2192–2217.\\n',\n",
       " '        ◦ Show context for reference 44 \\n',\n",
       " '            ▪ CAS\\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    45. 45.\\n',\n",
       " '       Hemingway H, Marmot M. Evidence based cardiology-Psychosocial factors in the aetiology and prognosis of coronary heart disease: systematic review of prospective cohort studies. Bmj 1999; 318: 1460–1467.\\n',\n",
       " '        ◦ Show context for reference 45 \\n',\n",
       " '            ▪ CAS\\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    46. 46.\\n',\n",
       " '       Lett HS, Blumenthal JA, Babyak MA, Catellier DJ, Carney RM, Berkman LF et al. Dimensions of social support and depression in patients at increased psychosocial risk recovering from myocardial infarction. Int J Behav Med 2009; 16: 248–258.\\n',\n",
       " '        ◦ Show context for reference 46 \\n',\n",
       " '            ▪ PubMed\\n',\n",
       " '            ▪ Article\\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " '    47. 47.\\n',\n",
       " '       LDAvis: A method for visualizing and interpreting topics.Workshop on Interactive Language Learning, Visualizaiton, and Interfaces. Association for Computational Linguistics, Baltimore, Maryland, USA, 2014.\\n',\n",
       " '        ◦ Show context for reference 47 \\n',\n",
       " '            ▪ Google Scholar\\n',\n",
       " 'Download references\\n',\n",
       " 'AcknowledgementsAcknowledgements\\n',\n",
       " 'THM is supported in part by a Stanley Center fellowship, and a NARSAD grant. RHP is supported in part by grants P50 MH106933 and R01MH106577 from the National Institute of Mental Health and by grant P50 MH106933 from the National Human Genome Research Institute.\\n',\n",
       " 'Author informationAuthor information\\n',\n",
       " 'Affiliations\\n',\n",
       " '            ▪ Center for Quantitative Health, Center for Human Genetic Research and Department of Psychiatry, Massachusetts General Hospital, Boston, MA, USA\\n',\n",
       " '        ◦ T H McCoy\\n',\n",
       " '        ◦ , V M Castro\\n',\n",
       " '        ◦ , L Snapper\\n',\n",
       " '        ◦ , K Hart\\n',\n",
       " '        ◦ \\xa0& R H Perlis\\n',\n",
       " '            ▪ Partners Research Information Systems and Computing, Partners HealthCare System, One Constitution Center, Boston, MA, USA\\n',\n",
       " '        ◦ V M Castro\\n',\n",
       " '            ▪ Cardiology Division, Massachusetts General Hospital and Harvard Clinical Research Institute, Boston, MA, USA\\n',\n",
       " '        ◦ J L Januzzi\\n',\n",
       " '            ▪ Department of Psychiatry, Massachusetts General Hospital, Boston, MA, USA\\n',\n",
       " '        ◦ J C Huffman\\n',\n",
       " 'Authors\\n',\n",
       " '            ▪ Search for T H McCoy in:\\n',\n",
       " '        ◦ Nature Research journals • \\n',\n",
       " '        ◦ PubMed • \\n',\n",
       " '        ◦ Google Scholar\\n',\n",
       " '            ▪ Search for V M Castro in:\\n',\n",
       " '        ◦ Nature Research journals • \\n',\n",
       " '        ◦ PubMed • \\n',\n",
       " '        ◦ Google Scholar\\n',\n",
       " '            ▪ Search for L Snapper in:\\n',\n",
       " '        ◦ Nature Research journals • \\n',\n",
       " '        ◦ PubMed • \\n',\n",
       " '        ◦ Google Scholar\\n',\n",
       " '            ▪ Search for K Hart in:\\n',\n",
       " '        ◦ Nature Research journals • \\n',\n",
       " '        ◦ PubMed • \\n',\n",
       " '        ◦ Google Scholar\\n',\n",
       " '            ▪ Search for J L Januzzi in:\\n',\n",
       " '        ◦ Nature Research journals • \\n',\n",
       " '        ◦ PubMed • \\n',\n",
       " '        ◦ Google Scholar\\n',\n",
       " '            ▪ Search for J C Huffman in:\\n',\n",
       " '        ◦ Nature Research journals • \\n',\n",
       " '        ◦ PubMed • \\n',\n",
       " '        ◦ Google Scholar\\n',\n",
       " '            ▪ Search for R H Perlis in:\\n',\n",
       " '        ◦ Nature Research journals • \\n',\n",
       " '        ◦ PubMed • \\n',\n",
       " '        ◦ Google Scholar\\n',\n",
       " 'Competing interests\\n',\n",
       " 'RHP has served on advisory boards or provided consulting to Genomind, Healthrageous, Perfect Health, Pfizer, Psy Therapeutics, and RIDVentures. He receives patent royalties from Concordant Rater Systems (now Bracket). The remaining authors declare no conflict of interest.\\n',\n",
       " 'Corresponding author\\n',\n",
       " 'Correspondence to R H Perlis.\\n',\n",
       " 'Supplementary informationSupplementary information\\n',\n",
       " 'Excel files\\n',\n",
       " '    1. 1.\\n',\n",
       " '               Supplementary Table 1\\n',\n",
       " 'Image files\\n',\n",
       " '    1. 1.\\n',\n",
       " '               Supplementary Figure 1\\n',\n",
       " 'Rights and permissionsRights and permissions\\n',\n",
       " 'This work is licensed under a Creative Commons Attribution 4.0 International License. The images or other third party material in this article are included in the article’s Creative Commons license, unless indicated otherwise in the credit line; if the material is not included under the Creative Commons license, users will need to obtain permission from the license holder to reproduce the material. To view a copy of this license, visit http://creativecommons.org/licenses/by/4.0/\\n',\n",
       " 'About this articleAbout this article\\n',\n",
       " 'Publication history\\n',\n",
       " 'Received\\n',\n",
       " '19 January 2017\\n',\n",
       " 'Revised\\n',\n",
       " '07 July 2017\\n',\n",
       " 'Accepted\\n',\n",
       " '14 July 2017\\n',\n",
       " 'Published\\n',\n",
       " '19 September 2017\\n',\n",
       " 'DOI\\n',\n",
       " 'https://doi.org/10.1038/tp.2017.201\\n',\n",
       " 'Share this article\\n',\n",
       " 'Anyone you share the following link with will be able to read this content:\\n',\n",
       " 'Get shareable link\\n',\n",
       " 'Subjects\\n',\n",
       " '    • Clinical genetics\\n',\n",
       " '    • Human behaviour \\n',\n",
       " 'Supplementary Information accompanies the paper on the Translational Psychiatry website (http://www.nature.com/tp)\\n',\n",
       " 'Article Tools\\n',\n",
       " 'Download PDF\\n',\n",
       " '3Altmetric\\n',\n",
       " '\\n',\n",
       " 'Article metrics \\n',\n",
       " 'sections \\n',\n",
       " 'figures \\n',\n",
       " 'references \\n',\n",
       " '    • Abstract \\n',\n",
       " '    • Introduction \\n',\n",
       " '    • Materials and methods \\n',\n",
       " '    • Results \\n',\n",
       " '    • Discussion \\n',\n",
       " '    • References \\n',\n",
       " '    • Acknowledgements \\n',\n",
       " '    • Author information \\n',\n",
       " '    • Supplementary information \\n',\n",
       " '    • Rights and permissions \\n',\n",
       " '    • About this article \\n',\n",
       " '    • Comments \\n',\n",
       " 'References\\n',\n",
       " '    1. Whooley MA, de Jonge P, Vittinghoff E, Otte C, Moos R, Carney RM et al. Depressive symptoms, health behaviors, and risk of cardiovascular events in patients with coronary heart disease. JAMA 2008; 300: 2379–2388.\\n',\n",
       " '            ▪ CAS \\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    2. Whooley MA. Depression and cardiovascular disease: healing the broken-hearted. JAMA 2006; 295: 2874–2881.\\n',\n",
       " '            ▪ CAS \\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    3. Ducat L, Philipson LH, Anderson BJ. The mental health comorbidities of diabetes. JAMA 2014; 312: 691–692.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    4. Golden SH, Lazo M, Carnethon M, Bertoni AG, Schreiner PJ, Diez Roux AV et al. Examining a bidirectional association between depressive symptoms and diabetes. JAMA 2008; 299: 2751–2759.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    5. Cohen BE, Martires KJ, Ho RS. Psoriasis and the risk of depression in the US population: National Health and Nutrition Examination Survey 2009–2012. JAMA Dermatol 2016; 152: 73–79.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    6. Girardi P, Pompili M, Innamorati M, Mancini M, Serafini G, Mazzarini L et al. Duloxetine in acute major depression: review of comparisons to placebo and standard antidepressants using dissimilar methods. Hum Psychopharmacol 2009; 24: 177–190.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    7. Blume J, Douglas SD, Evans DL. Immune suppression and immune activation in depression. Brain Behav Immun 2011; 25: 221–229.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    8. Penninx BW. Depression and cardiovascular disease: epidemiological evidence on their linking mechanisms. Neurosci Biobehav Rev 2017; 74(Part B): 277–286.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    9. Tang P-L, Wang H-H, Chou F-H. A systematic review and meta-analysis of demoralization and depression in patients with cancer. Psychosomatics 2015; 56: 634–643.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    10. Cheng J, Zhang J, Lu C, Wang L. Using optogenetics to translate the “inflammatory dialogue” between heart and brain in the context of stress. Neurosci Bull 2012; 28: 435–448.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    11. van Hecke O, Hocking LJ, Torrance N, Campbell A, Padmanabhan S, Porteous DJ et al. Chronic pain, depression and cardiovascular disease linked through a shared genetic predisposition: Analysis of a family-based cohort and twin study. PLoS One 2017; 12: e0170653.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    12. Scherrer JF, Xian H, Bucholz KK, Eisen SA, Lyons MJ, Goldberg J et al. A twin study of depression symptoms, hypertension, and heart disease in middle-aged men. Psychosom Med 2003; 65: 548–557.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    13. Kendler KS, Gardner CO, Fiske A, Gatz M. Major depression and coronary artery disease in the swedish twin registry: Phenotypic, genetic, and environmental sources of comorbidity. Arch Gen Psychiatry 2009; 66: 857–863.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    14. Smith IG, Parker G, Rourke P, Cvejic E, Vollmer-Conna U. Acute coronary syndrome and depression: A review of shared pathophysiological pathways. Aust N Z J Psychiatry 2015; 49: 994–1005.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    15. Hyde CL, Nagle MW, Tian C, Chen X, Paciga SA, Wendland JR et al. Identification of 15 genetic loci associated with risk of major depression in individuals of European descent. Nat Genet 2016; 48: 1031–1036.\\n',\n",
       " '            ▪ CAS \\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    16. Antony A, Tang YS, Khan RA, Biju MP, Xiao X, Li QJ et al. Translational upregulation of folate receptors is mediated by homocysteine via RNA-heterogeneous nuclear ribonucleoprotein E1 interactions. J Clin Invest 2004; 113: 285–301.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    17. Blei DM, Ng AY, Jordan MI. Latent dirichlet allocation. J Machine Learn Res 2003; 3: 993–1022.\\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    18. Blei DM. Surveying a suite of algorithms that offer a solution to managing large document archives. Commun ACM 2012; 55: 77–84.\\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    19. Denny JC, Ritchie MD, Basford MA, Pulley JM, Bastarache L, Brown-Gentry K et al. PheWAS: demonstrating the feasibility of a phenome-wide scan to discover gene–disease associations. Bioinformatics 2010; 26: 1205–1210.\\n',\n",
       " '            ▪ CAS \\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    20. Arun R, Suresh V, Veni Madhavan C, Narasimha Murthy M. On finding the natural number of topics with latent dirichlet allocation: Some observations.In:Advance in Knowledge Discovery and Data Mining, Hyderabad, India,2010,pp391–402.\\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    21. Cao J, Xia T, Li J, Zhang Y, Tang S. A density-based method for adaptive LDA model selection. Neurocomputing 2009; 72: 1775–1781.\\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    22. Deveaud R, SanJuan E, Bellot P. Accurate and effective latent concept modeling for ad hoc information retrieval. Document numérique 2014; 17: 61–84.\\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    23. Griffiths TL, Steyvers M. Finding scientific topics. Proc Natl Acad Sci USA 2004; 101(suppl 1): 5228–5235.\\n',\n",
       " '            ▪ CAS \\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    24. Rehurek R, Sojka PSoftware framework for topic modelling with large corpora. In:Proceedings of the LREC 2010 Workshop on New Challenges for NLP Frameworks. Citeseer: Valletta, Malta, 2010,pp45–50.\\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    25. Hoffman M, Blei D, Bach FOnline learning for latent dirichlet allocation In: Advances in Neural Information Processing Systems 2010,pp856–864.\\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " \"    26. Henn BM, Hon L, Macpherson JM, Eriksson N, Saxonov S, Pe'er I et al. Cryptic distant relatives are common in both isolated and cosmopolitan genetic samples. PLoS ONE 2012; 7: e34267.\\n\",\n",
       " '            ▪ CAS \\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    27. Fuchsberger C, Abecasis GR, Hinds DAMinimac2: faster genotype imputation. Bioinformatics 2015; 31: 782–784.\\n',\n",
       " '            ▪ CAS \\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    28. Fuchsberger C, Forer L, Schoenherr S, Das S, Abecasis G.Michigan Imputation Server. Available at https://imputationserver.sph.umich.edu/index.html#!pages/home(accessed on 21 December 2016.)\\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    29. Minimac3. Center for statistical genetics. Available athttp://genome.sph.umich.edu/wiki/Minimac3(accessed on 21 Deecember 2016.)\\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    30. Delaneau O, Marchini J, Zagury J-F. A linear complexity phasing method for thousands of genomes. Nature methods 2012; 9: 179–181.\\n',\n",
       " '            ▪ CAS \\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    31. Chang CC, Chow CC, Tellier LC, Vattikuti S, Purcell SM, Lee JJ. Second-generation PLINK: rising to the challenge of larger and richer datasets. GigaScience 2015; 4: 7.\\n',\n",
       " '            ▪ CAS \\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    32. Purcell S, Neale B, Todd-Brown K, Thomas L, Ferreira MA, Bender D et al. PLINK: a tool set for whole-genome association and population-based linkage analyses. Am J Hum Genet 2007; 81: 559–575.\\n',\n",
       " '            ▪ CAS \\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    33. Sankoh AJ, Huque MF, Dubey SD. Some comments on frequently used multiple endpoint adjustment methods in clinical trials. Stat Med 1997; 16: 2529–2542.\\n',\n",
       " '            ▪ CAS \\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " \"    34. Perneger TV. What's wrong with Bonferroni adjustments. BMJ 1998; 316: 1236–1238.\\n\",\n",
       " '            ▪ CAS \\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    35. Angermann CE, Gelbrich G, Störk S, Gunold H, Edelmann F, Wachter R et al. Effect of escitalopram on All-cause mortality and hospitalization in patients with heart failure and depression: The MOOD-HF randomized clinical trial. JAMA 2016; 315: 2683.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    36. Angermann C, Deckert J, Ertl G. Escitalopram and outcomes among patients with depression and heart failure-reply. JAMA 2016; 316: 1494.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    37. Ströhle A, Rieckmann N. Escitalopram and outcomes among patients with depression and heart failure. JAMA 2016; 316: 1494.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    38. Carney RM, Rich MW, Freedland KE, Saini J, Simeone C, Clark K. Major depressive disorder predicts cardiac events in patients with coronary artery disease. Psychosom Med 1988; 50: 627–633.\\n',\n",
       " '            ▪ CAS \\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    39. Van Melle JP, De Jonge P, Spijkerman TA, Tijssen JG, Ormel J, Van Veldhuisen DJ et al. Prognostic association of depression following myocardial infarction with mortality and cardiovascular events: a meta-analysis. Psychosom Med 2004; 66: 814–822.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    40. Lespérance F, Frasure-Smith N, Talajic M. Major depression before and after myocardial infarction: its nature and consequences. Psychosom Med 1996; 58: 99–110.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    41. Rumsfeld JS, Jones PG, Whooley MA, Sullivan MD, Pitt B, Weintraub WS et al. Depression predicts mortality and hospitalization in patients with myocardial infarction complicated by heart failure. Am Heart J 2005; 150: 961–967.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    42. Williams M. Platelets and depression in cardiovascular disease: a brief review of the current literature. World J Psychiatr 2012; 2: 114–123.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    43. Can MM, Guler G, Guler E, Ozveren O, Turan B, DiNicolantinio JJ et al. Enhanced platelet reactivity in pediatric depression: an observational study. Blood Coagul Fibrinolysis 2015; 26: 731–735.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    44. Rozanski A, Blumenthal JA, Kaplan J. Impact of psychological factors on the pathogenesis of cardiovascular disease and implications for therapy. Circulation 1999; 99: 2192–2217.\\n',\n",
       " '            ▪ CAS \\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    45. Hemingway H, Marmot M. Evidence based cardiology-Psychosocial factors in the aetiology and prognosis of coronary heart disease: systematic review of prospective cohort studies. Bmj 1999; 318: 1460–1467.\\n',\n",
       " '            ▪ CAS \\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    46. Lett HS, Blumenthal JA, Babyak MA, Catellier DJ, Carney RM, Berkman LF et al. Dimensions of social support and depression in patients at increased psychosocial risk recovering from myocardial infarction. Int J Behav Med 2009; 16: 248–258.\\n',\n",
       " '            ▪ PubMed \\n',\n",
       " '            ▪ Article \\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " '    47. LDAvis: A method for visualizing and interpreting topics.Workshop on Interactive Language Learning, Visualizaiton, and Interfaces. Association for Computational Linguistics, Baltimore, Maryland, USA, 2014.\\n',\n",
       " '            ▪ Google Scholar \\n',\n",
       " 'Figure 1 \\n',\n",
       " ' \\n',\n",
       " '    • View in article \\n',\n",
       " '    • Full size image \\n',\n",
       " 'Figure 2 \\n',\n",
       " ' \\n',\n",
       " '    • View in article \\n',\n",
       " '    • Full size image \\n',\n",
       " 'Figure 3 \\n',\n",
       " ' \\n',\n",
       " '    • View in article \\n',\n",
       " '    • Full size image \\n',\n",
       " ' \\n',\n",
       " 'Translational Psychiatry\\n',\n",
       " 'ISSN 2158-3188 (online)\\n',\n",
       " '\\n',\n",
       " 'nature.com sitemap\\n',\n",
       " 'For authors and referees\\n',\n",
       " '    • Guide to Authors \\n',\n",
       " '    • Guide to Referees \\n',\n",
       " '    • Editorial policies \\n',\n",
       " '    • Language editing \\n',\n",
       " '    • Scientific editing \\n',\n",
       " '    • Reprints & permissions \\n',\n",
       " 'For readers\\n',\n",
       " '    • Journals A-Z \\n',\n",
       " '    • Subject Pages \\n',\n",
       " '    • Protocol Exchange \\n',\n",
       " '    • Blogs \\n',\n",
       " '    • Podcasts \\n',\n",
       " '    • Webcasts \\n',\n",
       " '    • Subscriptions \\n',\n",
       " '    • RSS \\n',\n",
       " 'For librarians\\n',\n",
       " '    • Account administration \\n',\n",
       " '    • Site licenses \\n',\n",
       " '    • Catalogs \\n',\n",
       " '    • Print subscriptions \\n',\n",
       " '    • Pay per view \\n',\n",
       " '    • Open access \\n',\n",
       " '    • Promotion tools \\n',\n",
       " '    • Public interfaces \\n',\n",
       " '    • User guides \\n',\n",
       " '    • Library relations \\n',\n",
       " '    • Publisher collaborations \\n',\n",
       " 'Careers and events\\n',\n",
       " '    • Nature conferences \\n',\n",
       " '    • Nature events \\n',\n",
       " '    • Nature jobs.com \\n',\n",
       " 'Regional websites\\n',\n",
       " '    • Nature asia \\n',\n",
       " '    • Nature China \\n',\n",
       " '    • Nature India \\n',\n",
       " '    • Nature Japan \\n',\n",
       " '    • Nature Korea \\n',\n",
       " '    • Nature Middle East \\n',\n",
       " 'Nature Research\\n',\n",
       " '    • About us \\n',\n",
       " '    • Work for us \\n',\n",
       " '    • Help \\n',\n",
       " '    • Contact us \\n',\n",
       " '    • Privacy Policy \\n',\n",
       " '    • Use of cookies \\n',\n",
       " '    • Manage cookies \\n',\n",
       " '    • Legal notice \\n',\n",
       " '    • Accessibility statement \\n',\n",
       " '    • Terms & Conditions \\n',\n",
       " '    • Partners \\n',\n",
       " 'For advertisers\\n',\n",
       " '    • Advertising options \\n',\n",
       " '    • Journal audience \\n',\n",
       " '    • Event advertising \\n',\n",
       " '    • Job posts \\n',\n",
       " '    • Contact representative \\n',\n",
       " 'For press\\n',\n",
       " '    • Press site \\n',\n",
       " '    • Press room \\n',\n",
       " '    • Contact press office \\n',\n",
       " ' © 2018 Springer Nature Limited. All rights reserved. \\n',\n",
       " '  \\n']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: fill\n",
    "texts = read_texts(\"/data/txt/directory\")\n",
    "text = list(texts)[4]\n",
    "text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Thank you for visiting nature.  You are using a browser version with limited support for CSS.  To obtain the best experience, we recommend you use a more up to date browser (or turn off compatibility mode in Internet Explorer).  In the meantime, to ensure continued support, we are displaying the site without styles and JavaScript. ',\n",
       " \" Major depressive disorder frequently co-occurs with medical disorders, raising the possibility of shared genetic liability.  Recent identification of 15 novel genetic loci associated with depression allows direct investigation of this question.  In cohorts of individuals participating in biobanks at two academic medical centers, we calculated polygenic loading for risk loci reported to be associated with depression.  We then examined the association between such loading and 50 groups of clinical diagnoses, or topics, drawn from these patients' electronic health records, determined using a novel application of latent Dirichilet allocation.  Three topics showed experiment-wide association with the depression liability score; these included diagnostic groups representing greater prevalence of mood and anxiety disorders, greater prevalence of cardiac ischemia, and a decreased prevalence of heart failure.  The latter two associations persisted even among individuals with no mood disorder diagnosis.  This application of a novel method for grouping related diagnoses in biobanks indicate shared genetic risk for depression and cardiac disease, with a pattern suggesting greater ischemic risk and diminished heart failure risk. \",\n",
       " ' Major depression has been associated with a host of non-psychiatric comorbidities, ranging from cardiovascular disease to autoimmune disorders.  The impact of co-occurrence is often profound and bi-directional—that is, outcomes of each disorder tend to be worse in the presence of the other,1, 2, 3, 4, 5 despite the availability of multiple efficacious treatments. 6 Thus, understanding the mechanism of such co-occurrence has both scientific and clinical relevance. ',\n",
       " ' This co-occurrence has a range of possible explanations.  It is possible that depression increases risk for some disorders, either directly (for example, via changes in cortisol or immunosuppression) or indirectly (for example, via changes in health behaviors). 7, 8 Conversely, the presence of a medical disorder can represent a stressor increasing risk for depression. 9 Yet, a third model posits a shared liability—i. , the same factors that increase risk for one disorder may increase risk for the other. 10 Multiple twin or family investigations support this shared liability, at least for cardiovascular disease. 11, 12, 13 In reality, it is likely that all three mechanisms play a role in co-occurrence of some disorders, but confirming the presence of at least one mechanism would represent a key first step in understanding these relationships. 14',\n",
       " ' Multiple novel genetic loci associated with major depression in individuals of Northern European ancestry at a genome-wide level of significance have recently been reported. 15 These newly reported loci allow for direct examination of the possibility that aggregated genetic liability for depression, in addition to depression itself, is associated with risk for disorders other than depression.  The present study aims to test the association between such liability and medical comorbidity, using a novel method for deriving this comorbidity.  Specifically, we drew on these newly reported genetic associations to construct estimates of common-variant genetic loading in the large medical biobanks from two academic medical centers.  We then tested these depression loading measures for association with empirically derived groups of diagnoses, or topics, that tend to co-occur with each other. ',\n",
       " ' Standard phenome-wide association studies (PheWAS) test all diagnostic codes against all predictors—typically individual risk variants or genome-wide common variants. 16 This approach risks either inflation of type I error (by testing 1500+ diagnostic codes), or type II error (by correcting for 1500+ diagnostic codes).  Further, it fails to take into account correlation between individual coded diagnoses, and the highly variable reliability of many such codes. ',\n",
       " \" As an alternative, we developed a method that applies latent Dirichilet allocation (LDA) to reduce categorical diagnostic ontologies to a finite set of topics on which these codes load. 17, 18 This form of unsupervised machine learning has most commonly been applied in natural language processing to capture the topics expressed in documents; it presumes that individual tokens or terms (here, diagnostic codes) reflect an underlying topic, and that the record of an individual patient reflects a combination of latent topics.  Figure 1 depicts the process of conceptualizing patient medical records as a 'bag' of observed diagnostic code counts from which unobserved latent topics are inferred using LDA.  Thereafter, the inferred topics are treated as the phenotype in analysis. \",\n",
       " \" Here we extracted all ICD-9 diagnostic codes for biobank participants from the inpatient and outpatient electronic medical records of Massachusetts General Hospital and Brigham and Women's Hospital and grouped them into 1667 PheWAS disease categories. 19 All participants had signed written informed consent for biobank participation, including consent for release of deidentified data under a Data Use Agreement to qualified investigators, as approved by the hospitals' Institutional Review Board.  We eliminated PheWAS codes occurring in <1% of individuals or more than 99% of individuals in the first cohort, leaving 508 codes for topic model construction.  We then trained a 50-topic model, and scored cohorts one and two.  (The decision a priori to select 50 topics is discussed further below; the optimal number of targets remains an area of research in unsupervised learning20, 21, 22, 23).  The LDA was performed using the Gensim implementation. 24, 25 Importantly LDA allows for the possibility of all codes with respect to all topics.  The distinction between topics is in the expected probabilities of each code.  As such, we focus on the most probable diagnostic codes given each topic as a means of interpreting the topics.  When individual topics are mentioned in the text they are named using the most strongly loaded code with the suffix ‘++’ to indicate that a topic comprises many codes, each contributing to membership in that topic. \",\n",
       " ' All subjects were genotyped using either the Illumina MEGA (cohort 1, n=4931) or the Illumina MEGA-EX (cohort 2, n=4428) array (Illumina; San Diego, CA, USA).  Each cohort was cleaned, imputed, and analyzed separately to minimize batch effects.  We retained subjects with genotyping call rates exceeding 99% and no evidence of relatedness based on identity by descent (IBD). 26 We likewise retained any SNPs with call rate of 95% or greater, and Hardy–Weinberg equilibrium P-value>1 × 10−10.  Genotypes were next imputed using the Michigan Imputation Server implementing Minimac3, based on all population subsets from 1000 Genomes Phase 3 v5 as reference panel. 27, 28, 29 Phasing of haplotypes used SHAPEIT. 30',\n",
       " ' We generated principal components, as implemented in PLINK 1. 9, to identify the first 10 components of the variance-standardized relationship matrix among genotyped SNPs in each cohort. 31 After overlaying HapMap populations, only those individuals falling within the Northern European cluster were included in subsequent analysis. ',\n",
       " \" We generated polygenic risk scores (PRS), estimates of polygenic loading for MDD, using seven tranches of SNPs (5 × 10−8, 1 × 10−7, 1 × 10−6, 1 × 10−5, 1 × 10−4, 1 × 10−3 and 1 × 10−2, S1–S7) drawn from our prior publication reporting 15 genome-wide associations with depression. 15 The value for each P-value tranche represents the maximum P-value that is included in that tranche.  This list was linkage-disequilibrium pruned using the 'clump' function as implemented in PLINK 1. 9, with a 250\\u2009kb window and minimum r2 set at 0. 32\",\n",
       " ' We used linear regression to examine association between depression polygenic score and each of the 50 topics.  We fit both unadjusted models and models incorporating the first 10 MDS components, and present the meta-analyzed result of the two genotyping cohorts. ',\n",
       " ' As this analysis was intended as a hypothesis-generating effort, we present uncorrected P-values in all results.  For purposes of interpretation, Bonferroni correction for 50 topics would require a P-value of 0. 001, for significance.  The seven PRS tranches are correlated so do not represent seven independent tests per phenotype; since the average r between them is ~0. 65, we consider a fully corrected P-value threshold for significance to be 0. 33, 34 To further elucidate the statistical significance of any associations identified we utilized permutation to calculate empirical P-values as well as an experiment-wide P-value.  To do so, we randomized the relationship between the topics (phenotype) and the MDS adjusted PRS (genotype) and calculated association in the full cohort between all tranches and all topics 100\\u2009000 times under this simulated null. ',\n",
       " ' In the first genotyping cohort, there were 3728 individuals, including 2165 females (58. 1%) and 997 individuals (26. 7%) with a mood disorder.  Cohort 2 included 2712 individuals, including 712 (49. 2%) females and 779 individuals (28. 7%) with a mood disorder.  Mean age in cohort 1 was 57. 8) years; mean age in cohort 2 was 62. ',\n",
       " ' Figure 2 illustrates the distribution of associations with PRS by topic and minimum P-value (that is, PRS threshold yielding strongest evidence of association).  For consistency with other data clustering methods, topics are named according to predominant terms, adding the suffix ‘++’ to indicate that topics may contain overlapping terms as well as apparently unrelated terms.  Three topics—mood disorder++ (03), heart failure++ (21) and cardiac ischemia++ (27)—met an experiment-wide threshold for association (Table 1).  PheWAS diagnostic codes contributing to these topics are listed in Table 2, ranked by order of contribution (that is, weighting of each code for a given topic, from greatest to least).  (Supplementary Table 1 reports all topics by all PRS tranches, sorted by association P-value).  Permuted topic level associations matched those of the primary analysis (Supplementary Figure 1).  In experiment-wide permutation analysis the number of significant topics was itself statistically significant (permuted P=0. ',\n",
       " ' Table 1: Topics associated with PRS at an experiment-wide threshold for significance in primary analysis (left); follow-up analysis of primary associations (right)',\n",
       " ' In secondary analysis, we excluded any individuals with a mood disorder PheWAS code (n=1776) and repeated these analyses.  Association P-values are indicated in Table 1 (right), and visualized in a heat map in Figure 3.  As anticipated, the mood disorder topic (03) was no longer significant, suggesting that the additional codes in that topic do not contribute meaningfully to association; the two cardiac topics demonstrate persistent association. ',\n",
       " ' In this analysis of electronic health record data from 6440 individuals of Northern European ancestry, we identified three sets of diagnoses (topics) associated with PRS for depression at an experiment-wide significance threshold.  One of these encompasses mood disorders, and can be considered a positive control or indicator of assay sensitivity—though notably, it also includes related codes (adjustment disorder, tobacco use, and anxiety) that may reflect true pleiotropy or simply differences in the way mood and anxiety symptoms are coded.  The other two reflect different elements of cardiac pathology.  That cardiac pathology surfaces from this unsupervised machine learning approach is reassuringly face valid given the extensive literature relating cardiac and mood disorders; however, cardiac pathology is subdivided into distinct associations. 1, 2, 3, 4 Cardiac ischemia++ (topic 27) largely captures acute coronary syndromes, including risk factors for such syndromes (See Table 2 for the codes most strongly associated with this topic).  The data suggest that MDD loading is associated with greater risk for these acute syndromes (Table 1).  Heart failure++ (topic 21) also reflects cardiovascular pathology but appears to reflect a more chronic disease course, including heart failure but not limited to cardiac disease (Table 2).  Depressive genetic loading is inversely associated with heart failure++ (Table 1).  This differential result underscores the complexity of the relationship between depression and cardiovascular disease. 35, 36, 37 Importantly, evidence of association persists even when individuals with a mood disorder are excluded, suggesting that the observed associations are not simply consequences of a mood disorder diagnosis. ',\n",
       " ' The relationship between depressive disorder and acute coronary syndromes is well described; however, the mechanism remains unclear. 38, 39, 40, 41, 42 Multiple investigations using twin- and family-based designs have found evidence of shared heritable liability between MDD and cardiac disease. 11, 12, 13 One study suggests the complexity of this relationship, with the extent of liability varying with sex and age. 13 Further complicating efforts to understand comorbidity is the observation that depression is associated with increased deleterious behavior such as tobacco use and less exercise - notably tobacco use, for which risk is likely to be heritable, contributes to both a mood++ and a cardiovascular++ topic in our results.  On the other hand, in prior studies individuals with major depression exhibit greater platelet reactivity and prevalent endothelial dysfunction relative to non-depressed subjects, which is a candidate pathway for depression increasing cardiac risk directly. 43',\n",
       " ' Surprisingly, in the present study depression loading is inversely associated with more chronic forms of cardiac disease (heart failure++ topic 21).  Although it is speculative, the divergences between acute and chronic cardiac pathology may reflect the tension between genetic and behavioral determinates of health.  In the chronic phases of cardiac disease, personality is an important determinate of outcome. 44, 45, 46 Alternatively, those with heart failure and comorbid depression might in fact have substantially higher mortality risk, obscuring a chronic link between the two diagnoses due to attrition. ',\n",
       " ' In interpreting these associations, it is important to recognize that the methodology itself identifies co-occurring codes; the application of names to the underlying (that is, latent) concepts requires human intervention.  Moreover, codes are associated, even if only negligibly, with all topics.  This characteristic represents both a feature and a limitation of the methodology: it does not require manual curation of topics, and thus can discover relationships in data not specified a priori; on the other hand, it requires interpretation of results that may not be intuitive.  Concretely, we would not have expected acute and chronic cardiac disease to primarily occur in distinct topics.  This surprise bespeaks the hypothesis-generating potential of unsupervised machine learning, which still requires interpretation and follow-up.  For ease of reading, by convention we name each topic by the most strongly-loading code, but again emphasize that a topic is not simply a single code and note that visualization and interpretation of probabilistic topics remain an area of active investigation in machine learning. 47 Conversely, for testing specific hypotheses about a particular set of diagnoses, merely testing a curated code or set of codes would be most interpretable, but preclude discovery of new or unanticipated relationships. ',\n",
       " ' Latent Dirichilet allocation has been applied extensively in natural language processing to identify topics reflected in blocks of text; however, to our knowledge, it has not been utilized for genetic investigation.  The present study suggests the utility of this method for understanding the relationship between psychiatric and comorbid disorders, and we hope this study will prompt further investigations in larger biobanks and in using other polygenic risk measures.  Still, an important caveat to this work is the inability to determine causation.  Specifically, we cannot distinguish between the case of shared genetic liability for depression and comorbidities, and depression ‘causing’ the observed comorbidity.  However, we are able to render the third possibility (the comorbidity causes the depression) less likely, by conditioning on an aggregate measure of genetic depression liability.  A further caveat is the likelihood that other parameters would yield differing results—i. , the application of 10 or 100 topics.  Here, we selected 50 as a plausible number of disease groups a priori, and based on prior work with LDA in other contexts.  Understanding the extent to which other, nonlinear means of weighting diagnoses, or specifying other numbers of topics, would yield differing results merits further study. ',\n",
       " ' Taken together, our results suggest the complexity of the relationship between mood disorders and comorbid somatic illness, and indicate specific groups of diagnoses that may travel along with genetic risk for major depression.  They further illustrate the application of a novel approach to aggregating diagnoses, applicable to any large clinical data set, which may be more tolerant of heterogeneity in diagnostic codes and sensitive to groups of diagnoses that travel together.  At the same time, the observed association with mood disorders may be considered further replication of the previous report of depression liability genes. 15 The application of topic modeling therefore appears to be a promising strategy. ',\n",
       " '        Whooley MA, de Jonge P, Vittinghoff E, Otte C, Moos R, Carney RM et al.  Depressive symptoms, health behaviors, and risk of cardiovascular events in patients with coronary heart disease.  JAMA 2008; 300: 2379–2388. ',\n",
       " '        Golden SH, Lazo M, Carnethon M, Bertoni AG, Schreiner PJ, Diez Roux AV et al.  Examining a bidirectional association between depressive symptoms and diabetes.  JAMA 2008; 299: 2751–2759. ',\n",
       " '        Cohen BE, Martires KJ, Ho RS.  Psoriasis and the risk of depression in the US population: National Health and Nutrition Examination Survey 2009–2012.  JAMA Dermatol 2016; 152: 73–79. ',\n",
       " '        Girardi P, Pompili M, Innamorati M, Mancini M, Serafini G, Mazzarini L et al.  Duloxetine in acute major depression: review of comparisons to placebo and standard antidepressants using dissimilar methods.  Hum Psychopharmacol 2009; 24: 177–190. ',\n",
       " '        Tang P-L, Wang H-H, Chou F-H.  A systematic review and meta-analysis of demoralization and depression in patients with cancer.  Psychosomatics 2015; 56: 634–643. ',\n",
       " '        Cheng J, Zhang J, Lu C, Wang L.  Using optogenetics to translate the “inflammatory dialogue” between heart and brain in the context of stress.  Neurosci Bull 2012; 28: 435–448. ',\n",
       " '        van Hecke O, Hocking LJ, Torrance N, Campbell A, Padmanabhan S, Porteous DJ et al.  Chronic pain, depression and cardiovascular disease linked through a shared genetic predisposition: Analysis of a family-based cohort and twin study.  PLoS One 2017; 12: e0170653. ',\n",
       " '        Scherrer JF, Xian H, Bucholz KK, Eisen SA, Lyons MJ, Goldberg J et al.  A twin study of depression symptoms, hypertension, and heart disease in middle-aged men.  Psychosom Med 2003; 65: 548–557. ',\n",
       " '        Kendler KS, Gardner CO, Fiske A, Gatz M.  Major depression and coronary artery disease in the swedish twin registry: Phenotypic, genetic, and environmental sources of comorbidity.  Arch Gen Psychiatry 2009; 66: 857–863. ',\n",
       " '        Smith IG, Parker G, Rourke P, Cvejic E, Vollmer-Conna U.  Acute coronary syndrome and depression: A review of shared pathophysiological pathways.  Aust N Z J Psychiatry 2015; 49: 994–1005. ',\n",
       " '        Hyde CL, Nagle MW, Tian C, Chen X, Paciga SA, Wendland JR et al.  Identification of 15 genetic loci associated with risk of major depression in individuals of European descent.  Nat Genet 2016; 48: 1031–1036. ',\n",
       " '        Antony A, Tang YS, Khan RA, Biju MP, Xiao X, Li QJ et al.  Translational upregulation of folate receptors is mediated by homocysteine via RNA-heterogeneous nuclear ribonucleoprotein E1 interactions.  J Clin Invest 2004; 113: 285–301. ',\n",
       " '        Blei DM.  Surveying a suite of algorithms that offer a solution to managing large document archives.  Commun ACM 2012; 55: 77–84. ',\n",
       " '        Denny JC, Ritchie MD, Basford MA, Pulley JM, Bastarache L, Brown-Gentry K et al.  PheWAS: demonstrating the feasibility of a phenome-wide scan to discover gene–disease associations.  Bioinformatics 2010; 26: 1205–1210. ',\n",
       " '        Arun R, Suresh V, Veni Madhavan C, Narasimha Murthy M.  On finding the natural number of topics with latent dirichlet allocation: Some observations. In:Advance in Knowledge Discovery and Data Mining, Hyderabad, India,2010,pp391–402. ',\n",
       " '        Cao J, Xia T, Li J, Zhang Y, Tang S.  A density-based method for adaptive LDA model selection.  Neurocomputing 2009; 72: 1775–1781. ',\n",
       " '        Deveaud R, SanJuan E, Bellot P.  Accurate and effective latent concept modeling for ad hoc information retrieval.  Document numérique 2014; 17: 61–84. ',\n",
       " '        Rehurek R, Sojka PSoftware framework for topic modelling with large corpora.  In:Proceedings of the LREC 2010 Workshop on New Challenges for NLP Frameworks.  Citeseer: Valletta, Malta, 2010,pp45–50. ',\n",
       " \"        Henn BM, Hon L, Macpherson JM, Eriksson N, Saxonov S, Pe'er I et al.  Cryptic distant relatives are common in both isolated and cosmopolitan genetic samples.  PLoS ONE 2012; 7: e34267. \",\n",
       " '        Chang CC, Chow CC, Tellier LC, Vattikuti S, Purcell SM, Lee JJ.  Second-generation PLINK: rising to the challenge of larger and richer datasets.  GigaScience 2015; 4: 7. ',\n",
       " '        Purcell S, Neale B, Todd-Brown K, Thomas L, Ferreira MA, Bender D et al.  PLINK: a tool set for whole-genome association and population-based linkage analyses.  Am J Hum Genet 2007; 81: 559–575. ',\n",
       " '        Sankoh AJ, Huque MF, Dubey SD.  Some comments on frequently used multiple endpoint adjustment methods in clinical trials.  Stat Med 1997; 16: 2529–2542. ',\n",
       " '        Angermann CE, Gelbrich G, Störk S, Gunold H, Edelmann F, Wachter R et al.  Effect of escitalopram on All-cause mortality and hospitalization in patients with heart failure and depression: The MOOD-HF randomized clinical trial.  JAMA 2016; 315: 2683. ',\n",
       " '        Carney RM, Rich MW, Freedland KE, Saini J, Simeone C, Clark K.  Major depressive disorder predicts cardiac events in patients with coronary artery disease.  Psychosom Med 1988; 50: 627–633. ',\n",
       " '        Van Melle JP, De Jonge P, Spijkerman TA, Tijssen JG, Ormel J, Van Veldhuisen DJ et al.  Prognostic association of depression following myocardial infarction with mortality and cardiovascular events: a meta-analysis.  Psychosom Med 2004; 66: 814–822. ',\n",
       " '        Lespérance F, Frasure-Smith N, Talajic M.  Major depression before and after myocardial infarction: its nature and consequences.  Psychosom Med 1996; 58: 99–110. ',\n",
       " '        Rumsfeld JS, Jones PG, Whooley MA, Sullivan MD, Pitt B, Weintraub WS et al.  Depression predicts mortality and hospitalization in patients with myocardial infarction complicated by heart failure.  Am Heart J 2005; 150: 961–967. ',\n",
       " '        Williams M.  Platelets and depression in cardiovascular disease: a brief review of the current literature.  World J Psychiatr 2012; 2: 114–123. ',\n",
       " '        Can MM, Guler G, Guler E, Ozveren O, Turan B, DiNicolantinio JJ et al.  Enhanced platelet reactivity in pediatric depression: an observational study.  Blood Coagul Fibrinolysis 2015; 26: 731–735. ',\n",
       " '        Rozanski A, Blumenthal JA, Kaplan J.  Impact of psychological factors on the pathogenesis of cardiovascular disease and implications for therapy.  Circulation 1999; 99: 2192–2217. ',\n",
       " '        Hemingway H, Marmot M.  Evidence based cardiology-Psychosocial factors in the aetiology and prognosis of coronary heart disease: systematic review of prospective cohort studies.  Bmj 1999; 318: 1460–1467. ',\n",
       " '        Lett HS, Blumenthal JA, Babyak MA, Catellier DJ, Carney RM, Berkman LF et al.  Dimensions of social support and depression in patients at increased psychosocial risk recovering from myocardial infarction.  Int J Behav Med 2009; 16: 248–258. ',\n",
       " '        LDAvis: A method for visualizing and interpreting topics. Workshop on Interactive Language Learning, Visualizaiton, and Interfaces.  Association for Computational Linguistics, Baltimore, Maryland, USA, 2014. ',\n",
       " ' THM is supported in part by a Stanley Center fellowship, and a NARSAD grant.  RHP is supported in part by grants P50 MH106933 and R01MH106577 from the National Institute of Mental Health and by grant P50 MH106933 from the National Human Genome Research Institute. ',\n",
       " ' RHP has served on advisory boards or provided consulting to Genomind, Healthrageous, Perfect Health, Pfizer, Psy Therapeutics, and RIDVentures.  He receives patent royalties from Concordant Rater Systems (now Bracket).  The remaining authors declare no conflict of interest. ',\n",
       " \" This work is licensed under a Creative Commons Attribution 4. 0 International License.  The images or other third party material in this article are included in the article’s Creative Commons license, unless indicated otherwise in the credit line; if the material is not included under the Creative Commons license, users will need to obtain permission from the license holder to reproduce the material.  To view a copy of this license, visit http://creativecommons. org/licenses/by/4.  Whooley MA, de Jonge P, Vittinghoff E, Otte C, Moos R, Carney RM et al.  Depressive symptoms, health behaviors, and risk of cardiovascular events in patients with coronary heart disease.  JAMA 2008; 300: 2379–2388.  Golden SH, Lazo M, Carnethon M, Bertoni AG, Schreiner PJ, Diez Roux AV et al.  Examining a bidirectional association between depressive symptoms and diabetes.  JAMA 2008; 299: 2751–2759.  Cohen BE, Martires KJ, Ho RS.  Psoriasis and the risk of depression in the US population: National Health and Nutrition Examination Survey 2009–2012.  JAMA Dermatol 2016; 152: 73–79.  Girardi P, Pompili M, Innamorati M, Mancini M, Serafini G, Mazzarini L et al.  Duloxetine in acute major depression: review of comparisons to placebo and standard antidepressants using dissimilar methods.  Hum Psychopharmacol 2009; 24: 177–190.  Tang P-L, Wang H-H, Chou F-H.  A systematic review and meta-analysis of demoralization and depression in patients with cancer.  Psychosomatics 2015; 56: 634–643.  Cheng J, Zhang J, Lu C, Wang L.  Using optogenetics to translate the “inflammatory dialogue” between heart and brain in the context of stress.  Neurosci Bull 2012; 28: 435–448.  van Hecke O, Hocking LJ, Torrance N, Campbell A, Padmanabhan S, Porteous DJ et al.  Chronic pain, depression and cardiovascular disease linked through a shared genetic predisposition: Analysis of a family-based cohort and twin study.  PLoS One 2017; 12: e0170653.  Scherrer JF, Xian H, Bucholz KK, Eisen SA, Lyons MJ, Goldberg J et al.  A twin study of depression symptoms, hypertension, and heart disease in middle-aged men.  Psychosom Med 2003; 65: 548–557.  Kendler KS, Gardner CO, Fiske A, Gatz M.  Major depression and coronary artery disease in the swedish twin registry: Phenotypic, genetic, and environmental sources of comorbidity.  Arch Gen Psychiatry 2009; 66: 857–863.  Smith IG, Parker G, Rourke P, Cvejic E, Vollmer-Conna U.  Acute coronary syndrome and depression: A review of shared pathophysiological pathways.  Aust N Z J Psychiatry 2015; 49: 994–1005.  Hyde CL, Nagle MW, Tian C, Chen X, Paciga SA, Wendland JR et al.  Identification of 15 genetic loci associated with risk of major depression in individuals of European descent.  Nat Genet 2016; 48: 1031–1036.  Antony A, Tang YS, Khan RA, Biju MP, Xiao X, Li QJ et al.  Translational upregulation of folate receptors is mediated by homocysteine via RNA-heterogeneous nuclear ribonucleoprotein E1 interactions.  J Clin Invest 2004; 113: 285–301.  Surveying a suite of algorithms that offer a solution to managing large document archives.  Commun ACM 2012; 55: 77–84.  Denny JC, Ritchie MD, Basford MA, Pulley JM, Bastarache L, Brown-Gentry K et al.  PheWAS: demonstrating the feasibility of a phenome-wide scan to discover gene–disease associations.  Bioinformatics 2010; 26: 1205–1210.  Arun R, Suresh V, Veni Madhavan C, Narasimha Murthy M.  On finding the natural number of topics with latent dirichlet allocation: Some observations. In:Advance in Knowledge Discovery and Data Mining, Hyderabad, India,2010,pp391–402.  Cao J, Xia T, Li J, Zhang Y, Tang S.  A density-based method for adaptive LDA model selection.  Neurocomputing 2009; 72: 1775–1781.  Deveaud R, SanJuan E, Bellot P.  Accurate and effective latent concept modeling for ad hoc information retrieval.  Document numérique 2014; 17: 61–84.  Rehurek R, Sojka PSoftware framework for topic modelling with large corpora.  In:Proceedings of the LREC 2010 Workshop on New Challenges for NLP Frameworks.  Citeseer: Valletta, Malta, 2010,pp45–50.  Henn BM, Hon L, Macpherson JM, Eriksson N, Saxonov S, Pe'er I et al.  Cryptic distant relatives are common in both isolated and cosmopolitan genetic samples.  PLoS ONE 2012; 7: e34267.  Delaneau O, Marchini J, Zagury J-F.  A linear complexity phasing method for thousands of genomes.  Nature methods 2012; 9: 179–181.  Chang CC, Chow CC, Tellier LC, Vattikuti S, Purcell SM, Lee JJ.  Second-generation PLINK: rising to the challenge of larger and richer datasets.  GigaScience 2015; 4: 7.  Purcell S, Neale B, Todd-Brown K, Thomas L, Ferreira MA, Bender D et al.  PLINK: a tool set for whole-genome association and population-based linkage analyses.  Am J Hum Genet 2007; 81: 559–575.  Sankoh AJ, Huque MF, Dubey SD.  Some comments on frequently used multiple endpoint adjustment methods in clinical trials.  Stat Med 1997; 16: 2529–2542.  Angermann CE, Gelbrich G, Störk S, Gunold H, Edelmann F, Wachter R et al.  Effect of escitalopram on All-cause mortality and hospitalization in patients with heart failure and depression: The MOOD-HF randomized clinical trial.  JAMA 2016; 315: 2683.  Angermann C, Deckert J, Ertl G.  Escitalopram and outcomes among patients with depression and heart failure-reply.  JAMA 2016; 316: 1494.  Carney RM, Rich MW, Freedland KE, Saini J, Simeone C, Clark K.  Major depressive disorder predicts cardiac events in patients with coronary artery disease.  Psychosom Med 1988; 50: 627–633.  Van Melle JP, De Jonge P, Spijkerman TA, Tijssen JG, Ormel J, Van Veldhuisen DJ et al.  Prognostic association of depression following myocardial infarction with mortality and cardiovascular events: a meta-analysis.  Psychosom Med 2004; 66: 814–822.  Lespérance F, Frasure-Smith N, Talajic M.  Major depression before and after myocardial infarction: its nature and consequences.  Psychosom Med 1996; 58: 99–110.  Rumsfeld JS, Jones PG, Whooley MA, Sullivan MD, Pitt B, Weintraub WS et al.  Depression predicts mortality and hospitalization in patients with myocardial infarction complicated by heart failure.  Am Heart J 2005; 150: 961–967.  Platelets and depression in cardiovascular disease: a brief review of the current literature.  World J Psychiatr 2012; 2: 114–123.  Can MM, Guler G, Guler E, Ozveren O, Turan B, DiNicolantinio JJ et al.  Enhanced platelet reactivity in pediatric depression: an observational study.  Blood Coagul Fibrinolysis 2015; 26: 731–735.  Rozanski A, Blumenthal JA, Kaplan J.  Impact of psychological factors on the pathogenesis of cardiovascular disease and implications for therapy.  Circulation 1999; 99: 2192–2217.  Hemingway H, Marmot M.  Evidence based cardiology-Psychosocial factors in the aetiology and prognosis of coronary heart disease: systematic review of prospective cohort studies.  Bmj 1999; 318: 1460–1467.  Lett HS, Blumenthal JA, Babyak MA, Catellier DJ, Carney RM, Berkman LF et al.  Dimensions of social support and depression in patients at increased psychosocial risk recovering from myocardial infarction.  Int J Behav Med 2009; 16: 248–258.  LDAvis: A method for visualizing and interpreting topics. Workshop on Interactive Language Learning, Visualizaiton, and Interfaces.  Association for Computational Linguistics, Baltimore, Maryland, USA, 2014\"]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text_f = filter_sentences_from_text(text)\n",
    "text_f.split(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "texts_preproc = get_texts_from_dir(\"/data/gensim_citations_bkp/merged\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "texts = [t for t in texts_preproc.values() if len(t) > 0]\n",
    "texts_links = texts_preproc.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "dictionary = Dictionary(texts)\n",
    "corpus = [dictionary.doc2bow(text) for text in texts]\n",
    "lda = None\n",
    "\n",
    "\n",
    "def lda_from_texts(corpus, num_topics=20):\n",
    "    lda = LdaModel(corpus, num_topics=num_topics, alpha='auto', eval_every=5)\n",
    "    return lda"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "740"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "texts.__len__()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def topic_distro_for_text(text):\n",
    "    return lda.get_document_topics(dictionary.doc2bow(text), minimum_probability=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def terms_for_topic(topic_id, top_terms=10):\n",
    "    topic_top_terms = lda.get_topic_terms(topic_id, topn=top_terms)\n",
    "    return [dictionary.get(term[0]) for term in topic_top_terms]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### tf-idf integration\n",
    "\n",
    "Two approaches:\n",
    "\n",
    "1. use idf weights instead of BoW frequencies\n",
    "2. filter out given percentile of least important words from Docs' representation\n",
    "3. combined"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. approach"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tfidf = TfidfModel(corpus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "def terms_for_doc(doc_id):\n",
    "    return tfidf[corpus[doc_id]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "def top_terms_idf_for_doc(doc_id, percentile):\n",
    "    doc_terms_ordered = sorted(terms_for_doc(doc_id), key=lambda term: term[1],  reverse=True)\n",
    "    return [term[0] for term in doc_terms_ordered[:int(len(doc_terms_ordered)*percentile)]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "def name_terms_in_tuples(doc_corpus):\n",
    "    return [(dictionary.get(tup[0]), tup[1]) for tup in doc_corpus]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "def top_terms_idf_for_doc(doc_id, percentile):\n",
    "    doc_terms_ordered = sorted(terms_for_doc(doc_id), key=lambda term: term[1], reverse=True)\n",
    "    return [(doc_terms_ordered[i][0], doc_terms_ordered[i][1]) \n",
    "            for i in range(int(len(doc_terms_ordered)*percentile))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "def term_tfidf_for_doc(doc_idx):\n",
    "    return [(tfidf_term_tuple[0], tfidf_term_tuple[1]) \n",
    "            for tfidf_term_tuple in tfidf_corpus[doc_idx]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "tfidf_corpus1 = [top_terms_idf_for_doc(doc_i, 0.5) for doc_i in range(len(texts))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(8, 0.4263888832753634),\n",
       " (21, 0.40919533313685524),\n",
       " (24, 0.21344281382618024),\n",
       " (54, 0.18688720913165416),\n",
       " (39, 0.17642169230132967),\n",
       " (75, 0.17191612815653237),\n",
       " (46, 0.16980827584756597),\n",
       " (77, 0.16360395928964705),\n",
       " (63, 0.16323908336465176),\n",
       " (44, 0.14643919047125745),\n",
       " (34, 0.13978815535186268),\n",
       " (76, 0.13918626482458266),\n",
       " (17, 0.1375865905197532),\n",
       " (33, 0.1315216872647399),\n",
       " (7, 0.12374860032798168),\n",
       " (68, 0.12374860032798168),\n",
       " (55, 0.11923367556056216),\n",
       " (6, 0.11210965283447827),\n",
       " (9, 0.11143296654456712),\n",
       " (30, 0.11143296654456712),\n",
       " (65, 0.10835576248491846),\n",
       " (12, 0.10422361531271052),\n",
       " (42, 0.10422361531271052),\n",
       " (45, 0.10422361531271052),\n",
       " (0, 0.09344360456582708),\n",
       " (29, 0.09202816064986685),\n",
       " (74, 0.09068926556411135),\n",
       " (31, 0.0894190626912025),\n",
       " (35, 0.08198741455756116),\n",
       " (56, 0.08198741455756116),\n",
       " (59, 0.08108682342458094),\n",
       " (58, 0.08021783858415205),\n",
       " (10, 0.07556100209969478),\n",
       " (47, 0.07462656012219744),\n",
       " (70, 0.07418489659844453),\n",
       " (38, 0.07352449674572176),\n",
       " (53, 0.07284695357154429),\n",
       " (18, 0.0722542938728129),\n",
       " (51, 0.0722542938728129)]"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfidf_corpus1[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. approach:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0, 0.09344360456582708),\n",
       " (1, 0.02851096836158827),\n",
       " (2, 0.02717207327583276),\n",
       " (3, 0.06262975130624464),\n",
       " (4, 0.022358188818841874),\n",
       " (5, 0.04883172353650697),\n",
       " (6, 0.11210965283447827),\n",
       " (7, 0.12374860032798168),\n",
       " (8, 0.4263888832753634),\n",
       " (9, 0.11143296654456712),\n",
       " (10, 0.07556100209969478),\n",
       " (11, 0.04765787692078198),\n",
       " (12, 0.10422361531271052),\n",
       " (13, 0.06023140803970311),\n",
       " (14, 0.04588830094737287),\n",
       " (15, 0.05508952505442333),\n",
       " (16, 0.044223567067140424),\n",
       " (17, 0.1375865905197532),\n",
       " (18, 0.0722542938728129),\n",
       " (19, 0.02099324057250092),\n",
       " (20, 0.0687932952598766),\n",
       " (21, 0.40919533313685524),\n",
       " (22, 0.034329537636779185),\n",
       " (23, 0.06786237626523418),\n",
       " (24, 0.21344281382618024),\n",
       " (25, 0.06575722936458926),\n",
       " (26, 0.06989407767593134),\n",
       " (27, 0.06348375065341606),\n",
       " (28, 0.026873113586261408),\n",
       " (29, 0.09202816064986685),\n",
       " (30, 0.11143296654456712),\n",
       " (31, 0.0894190626912025),\n",
       " (32, 0.0587525491675709),\n",
       " (33, 0.1315216872647399),\n",
       " (34, 0.13978815535186268),\n",
       " (35, 0.08198741455756116),\n",
       " (36, 0.06482264573917157),\n",
       " (37, 0.06023140803970311),\n",
       " (38, 0.07352449674572176),\n",
       " (39, 0.17642169230132967),\n",
       " (40, 0.042501667335824445),\n",
       " (41, 0.04364480229658334),\n",
       " (42, 0.10422361531271052),\n",
       " (43, 0.018969072772641243),\n",
       " (44, 0.14643919047125745),\n",
       " (45, 0.10422361531271052),\n",
       " (46, 0.16980827584756597),\n",
       " (47, 0.07462656012219744),\n",
       " (48, 0.0693375705065911),\n",
       " (49, 0.02937239528135656),\n",
       " (50, 0.07104607733227525),\n",
       " (51, 0.0722542938728129),\n",
       " (52, 0.0722542938728129),\n",
       " (53, 0.07284695357154429),\n",
       " (54, 0.18688720913165416),\n",
       " (55, 0.11923367556056216),\n",
       " (56, 0.08198741455756116),\n",
       " (57, 0.062213547780507215),\n",
       " (58, 0.08021783858415205),\n",
       " (59, 0.08108682342458094),\n",
       " (60, 0.026382730693466448),\n",
       " (61, 0.06023140803970311),\n",
       " (62, 0.050833906602445994),\n",
       " (63, 0.16323908336465176),\n",
       " (64, 0.05031798113198441),\n",
       " (65, 0.10835576248491846),\n",
       " (66, 0.026577720101329683),\n",
       " (67, 0.041231464462915604),\n",
       " (68, 0.12374860032798168),\n",
       " (69, 0.0716428176978779),\n",
       " (70, 0.07418489659844453),\n",
       " (71, 0.04765787692078198),\n",
       " (72, 0.025188266111853246),\n",
       " (73, 0.07104607733227525),\n",
       " (74, 0.09068926556411135),\n",
       " (75, 0.17191612815653237),\n",
       " (76, 0.13918626482458266),\n",
       " (77, 0.16360395928964705)]"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "terms_for_doc(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [],
   "source": [
    "def top_terms_freqs_for_doc(doc_id, percentile):\n",
    "    doc_terms_idf = dict(terms_for_doc(doc_id))\n",
    "    doc_terms_count = corpus[doc_id]\n",
    "    doc_terms_count_ordered = sorted(doc_terms_count, \n",
    "                                     key=lambda term_count: doc_terms_idf[term_count[0]], reverse=True)\n",
    "    return [(doc_terms_count_ordered[i][0], doc_terms_count_ordered[i][1])\n",
    "            for i in range(int(len(doc_terms_count_ordered)*percentile))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(8, 8),\n",
       " (21, 3),\n",
       " (24, 5),\n",
       " (54, 2),\n",
       " (39, 2),\n",
       " (75, 2),\n",
       " (46, 2),\n",
       " (77, 1),\n",
       " (63, 5),\n",
       " (44, 1),\n",
       " (34, 2),\n",
       " (76, 6),\n",
       " (17, 2),\n",
       " (33, 3),\n",
       " (7, 1),\n",
       " (68, 1),\n",
       " (55, 1),\n",
       " (6, 1),\n",
       " (9, 2),\n",
       " (30, 2),\n",
       " (65, 2),\n",
       " (12, 1),\n",
       " (42, 1),\n",
       " (45, 1),\n",
       " (0, 1),\n",
       " (29, 1),\n",
       " (74, 1),\n",
       " (31, 1),\n",
       " (35, 1),\n",
       " (56, 1),\n",
       " (59, 1),\n",
       " (58, 1),\n",
       " (10, 1),\n",
       " (47, 2),\n",
       " (70, 1),\n",
       " (38, 1),\n",
       " (53, 2),\n",
       " (18, 1),\n",
       " (51, 1)]"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfidf_corpus2 = [top_terms_freqs_for_doc(doc_i, 0.5) for doc_i in range(len(texts))]\n",
    "tfidf_corpus2[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### LDA computation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/michal/miniconda3/envs/doc_analyzer/lib/python3.6/site-packages/gensim/models/ldamodel.py:582: RuntimeWarning:\n",
      "\n",
      "overflow encountered in exp2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "lda_num_topics = 400\n",
    "lda = lda_from_texts(tfidf_corpus1, num_topics=lda_num_topics)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### follows visualization part - inherited from other project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# [terms_for_topic(i, top_terms=20) for i in (1, 4, 16, 20)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [],
   "source": [
    "topic_distros = np.array([topic_distro_for_text(text) for text in texts])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [],
   "source": [
    "topic_distros = topic_distros[:, :, 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_topic_docs_distros = np.identity(lda_num_topics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [],
   "source": [
    "topic_distros_based = np.append(topic_distros, base_topic_docs_distros) \\\n",
    "    .reshape((len(topic_distros)+lda_num_topics, lda_num_topics))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1140, 400)"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topic_distros_based.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get a projection of topic according to their relative similarity\n",
    "\n",
    "Relative similarity is a correlation of documents' belonging to it"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get a projection of docs so that the docs are close to their major topic\n",
    "\n",
    "Doc distance to a cluster must be proportional to it's probability of belonging to it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.        , 0.94021965, 0.91278131, ..., 1.00638188, 1.00638193,\n",
       "        1.00638188],\n",
       "       [0.94021965, 0.        , 0.97272928, ..., 1.00847029, 1.00847033,\n",
       "        1.00847029],\n",
       "       [0.91278131, 0.97272928, 0.        , ..., 1.00876573, 1.00876599,\n",
       "        1.00876572],\n",
       "       ...,\n",
       "       [1.00638188, 1.00847029, 1.00876573, ..., 0.        , 1.00250627,\n",
       "        1.00250627],\n",
       "       [1.00638193, 1.00847033, 1.00876599, ..., 1.00250627, 0.        ,\n",
       "        1.00250627],\n",
       "       [1.00638188, 1.00847029, 1.00876572, ..., 1.00250627, 1.00250627,\n",
       "        0.        ]])"
      ]
     },
     "execution_count": 236,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from scipy.spatial.distance import pdist, squareform\n",
    "\n",
    "dists = squareform(pdist(topic_distros_based, metric=\"correlation\"))\n",
    "dists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1140, 1140)"
      ]
     },
     "execution_count": 237,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dists.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn import manifold\n",
    "\n",
    "adist = dists\n",
    "\n",
    "amax = np.amax(adist)\n",
    "adist /= amax\n",
    "\n",
    "mds = manifold.MDS(n_components=2, dissimilarity=\"precomputed\", random_state=6)\n",
    "results = mds.fit(adist)\n",
    "\n",
    "coords = results.embedding_\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly\n",
    "plotly.tools.set_credentials_file(username='stmichal', api_key='OXox9Rf8jzEHqUsNPqwn')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "\n",
    "all_topics_w = [terms_for_topic(i, top_terms=100) for i in range(lda_num_topics)]\n",
    "all_words = reduce(lambda x, y: set(x) | set(y), all_topics_w)\n",
    "intersect_words = list(filter(lambda w: sum([w in t_words for t_words in all_topics_w]) > 5, all_words))\n",
    "unique_topics_w = [[w for w in t_words if w not in intersect_words] for t_words in all_topics_w]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe id=\"igraph\" scrolling=\"no\" style=\"border:none;\" seamless=\"seamless\" src=\"https://plot.ly/~stmichal/81.embed\" height=\"1000px\" width=\"1000px\"></iframe>"
      ],
      "text/plain": [
       "<plotly.tools.PlotlyDisplay object>"
      ]
     },
     "execution_count": 243,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import plotly.plotly as py\n",
    "import plotly.graph_objs as go\n",
    "\n",
    "docs_len = len(topic_distros)\n",
    "\n",
    "# Create a trace\n",
    "trace_docs = go.Scatter(\n",
    "    x = coords[:docs_len, 0],\n",
    "    y = coords[:docs_len, 1],\n",
    "    mode = 'markers',\n",
    "    text = list(texts_links)\n",
    ")\n",
    "trace_bases = go.Scatter(\n",
    "    x = coords[docs_len:, 0],\n",
    "    y = coords[docs_len:, 1],\n",
    "    mode = 'markers',\n",
    "    marker = dict(color = 'rgba(255, 0, 122, .2)', size = 30),\n",
    "    text = [\"T %s: %s\" % (i, unique_topics_w[i]) for i in range(lda_num_topics)]\n",
    ")\n",
    "\n",
    "data = [trace_docs, trace_bases]\n",
    "\n",
    "label = 'MDS over LDA %s topics. tfidf for top 0.5 terms as frequencies' % lda.num_topics\n",
    "\n",
    "# Plot and embed in ipython notebook!\n",
    "layout = dict(title=label,  \n",
    "              font=dict(size=12),\n",
    "              showlegend=True,\n",
    "              width=1000,\n",
    "              height=1000,\n",
    "              margin=dict(l=40, r=40, b=85, t=100),\n",
    "              hovermode='closest',\n",
    "              plot_bgcolor='rgb(256,256,256)'          \n",
    "              )\n",
    "py.iplot(dict(data=data, layout=layout), filename=label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
